### PR TITLE
[feat] Add bitwise rotate operators for const bits

### PR DESCRIPTION
**Pull request by nyunyunyunyu**
- Add `const_right_rotate_unsafe` and `const_left_rotate_unsafe` into `GateInstructions`. 
- These functions can bitwise left/rotate rotate a unsigned integer by a const bit.
- These functions don't do range checking so it's unsafe.

### CODE REVIEW

**Review by jonathanpwang — CHANGES_REQUESTED** [review]
Some questions about shift indices, otherwise looks good.

Can you also please run `rustfmt` on all docs?

You can do this automatically in vscode with setting:
```
  "editor.formatOnSave": true,
```

**Comment by jonathanpwang** [comment]
Also please fix merge issues (merge community-edition updates)

File: halo2-base/src/gates/flex_gate.rs [hunk 1]
```diff
@@ -885,6 +907,39 @@ impl<F: ScalarField> GateChip<F> {
...
+            l_witness,
+            Constant(self.pow_of_two()[bit]),
+            r_witness,
+        );
+        ctx.constrain_equal(&a, &val_witness);
+        // Return (r_witness << (num_bits - bit + 1)) | l_witness
+        self.mul_add(
+            ctx, 
+            r_witness,
+            Constant(self.pow_of_two()[num_bits - bit + 1]),
```
[the remark this answers is not in the document]
Comment (nyunyunyunyu) [reply halo2-base/src/gates/flex_gate.rs]: Yes you are right. It was a dumb bug :(

File: halo2-base/src/gates/tests/bitwise_rotate.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#[test]
+fn test_bitwise_rotate() {
+    // 1 << 8 == 256
+    test_bitwise_rotate_gen::<8, 10>(8, true, 1, 256, true);
+    // 1 << 8 != 255
+    test_bitwise_rotate_gen::<8, 10>(8, true, 1, 255, false);
+    // 1 << 8 != 1
+    test_bitwise_rotate_gen::<8, 10>(8, true, 1, 1, false);
+    // 1 >> 1 == 1024
+    test_bitwise_rotate_gen::<1, 10>(8, false, 1, 1024, true);
```
[the remark this answers is not in the document]
Comment (nyunyunyunyu) [reply halo2-base/src/gates/tests/bitwise_rotate.rs]: It was a dumb bug :( Fixed

**Comment by nyunyunyunyu** [comment]
@PatStiles @jonathanpwang I think the naming in [halo2-base/src/gates/tests](https://github.com/axiom-crypto/halo2-lib/tree/community-edition/halo2-base/src/gates/tests) is inconsistent. Are you good if I:
1. Remove all "_test" suffixes? It's redundant because the folder name already contains "tests"
2. For testing utility functions(e.g. `tests_ground_truth.rs`), move all of them into `mod.rs` or `testing_utils.rs`?

**Review by jonathanpwang — APPROVED** [review]
LGTM!

**Review by jonathanpwang — APPROVED** [review]
LGTM!

**Comment by jonathanpwang** [comment]
> @PatStiles @jonathanpwang I think the naming in [halo2-base/src/gates/tests](https://github.com/axiom-crypto/halo2-lib/tree/community-edition/halo2-base/src/gates/tests) is inconsistent. Are you good if I:
> 
> 1. Remove all "_test" suffixes? It's redundant because the folder name already contains "tests"
> 2. For testing utility functions(e.g. `tests_ground_truth.rs`), move all of them into `mod.rs` or `testing_utils.rs`?

1. Sounds good
2. Sure, `testing_utils.rs` or just `utils.rs` are good.

**Comment by nyunyunyunyu** [comment]
> > @PatStiles @jonathanpwang I think the naming in [halo2-base/src/gates/tests](https://github.com/axiom-crypto/halo2-lib/tree/community-edition/halo2-base/src/gates/tests) is inconsistent. Are you good if I:
> > 
> > 1. Remove all "_test" suffixes? It's redundant because the folder name already contains "tests"
> > 2. For testing utility functions(e.g. `tests_ground_truth.rs`), move all of them into `mod.rs` or `testing_utils.rs`?
> 
> 1. Sounds good
> 2. Sure, `testing_utils.rs` or just `utils.rs` are good.

Great! I will do it in a sperate PR. 
I saw test bn254::tests::fixed_base_msm::bench_fixed_base_msm failed. I believe this PR doesn't touch halo2_ecc. Is the test flaky?

**Comment by jonathanpwang** [comment]
> > > @PatStiles @jonathanpwang I think the naming in [halo2-base/src/gates/tests](https://github.com/axiom-crypto/halo2-lib/tree/community-edition/halo2-base/src/gates/tests) is inconsistent. Are you good if I:
> > > 
> > > 1. Remove all "_test" suffixes? It's redundant because the folder name already contains "tests"
> > > 2. For testing utility functions(e.g. `tests_ground_truth.rs`), move all of them into `mod.rs` or `testing_utils.rs`?
> > 
> > 
> > 
> > 1. Sounds good
> > 2. Sure, `testing_utils.rs` or just `utils.rs` are good.
> 
> Great! I will do it in a sperate PR. I saw test bn254::tests::fixed_base_msm::bench_fixed_base_msm failed. I believe this PR doesn't touch halo2_ecc. Is the test flaky?

Yes I had to fix a typo: https://github.com/axiom-crypto/halo2-lib/commit/0b18ef13c69f21f6509f41e69bdcf047f0a8fedb

**Pull request merged**